### PR TITLE
dexclient read clientsecret from secret

### PR DESCRIPTION
**Pull request by cdoan1**
* update dexclient api to support secret, secret ref required
* update generate-cr.sh to create manifests example to test

### CODE REVIEW

File: hack/generate-cr.sh [hunk 1]
```diff
@@ -5,10 +5,23 @@ export NS=${NS:-"dex-operator"}
...
 export GITHUB_APP_CLIENTSECRET=${GITHUB_APP_CLIENTSECRET:-"githubappclientsecret"}
+
 export CLIENT_NAME=${CLIENT_NAME:-"thing"}
 export CLIENT_SECRET=${CLIENT_SECRET:-"thing123456"}
 export CLIENT_SECRET_NAME=cluster1-clientsecret
 
+export DEXSERVER_CLIENT_NAME=dexserverclientname
+export DEXSERVER_CLIENT_ID=dexserverclientid
+export DEXSERVER_CLIENT_SECRET_NAME=dexserversecret-secret
+export DEXSERVER_CLIENT_SECRET=dexserversecret123456
```
Comment (vidyanambiar) [inline hack/generate-cr.sh:16]: @cdoan1 Not seeing usages for these DEXSERVER_CLIENT_NAME, DEXSERVER_CLIENT_ID, DEXSERVER_CLIENT_SECRET_NAME, DEXSERVER_CLIENT_SECRET. Are they being used?

File: hack/generate-cr.sh [hunk 2]
```diff
@@ -68,24 +81,32 @@ spec:
       redirectURI: "https://${NAME}-${NS}.${APPS}/callback"
 EOF
 
-cat > demo-dexclient-hub.yaml <<EOF
```
Comment (vidyanambiar) [inline hack/generate-cr.sh:71]: The Readme still references this file. Could you update it please?

File: hack/generate-cr.sh [hunk 3]
```diff
@@ -68,24 +81,32 @@ spec:
...
   public: false
 EOF
 
 oc create secret generic ${CLIENT_SECRET_NAME} \
 --from-literal=clientSecret=${CLIENT_SECRET} \
--n openshift-config --dry-run -o yaml > demo-${CLIENT_SECRET_NAME}.yaml
+-n openshift-config --dry-run -o yaml > demo-oauth.yaml
 
-cat > demo-oauth.yaml <<EOF
+cat >> demo-oauth.yaml <<EOF
```
Comment (vidyanambiar) [inline hack/generate-cr.sh:108]: Should the clientID and clientSecret referenced below in demo-oauth be DEXCLIENT_ID and DEXCLIENT_SECRET_NAME?

**Comment by vidyanambiar** [comment]
I updated and applied the demo-oauth.yaml to reference the new DEXCLIENT_NAME, DEXCLIENT_ID and DEXCLIENT_SECRET_NAME variables. I may be missing something but when I apply the client and Oauth yaml and try logging in on the cluster with dex, I get this error:
image

File: api/v1alpha1/dexclient_types.go [hunk 4]
```diff
@@ -28,10 +29,9 @@ type DexClientSpec struct {
 	// +kubebuilder:validation:MinLength=4
 	// The name of the oidc config
 	ClientID string `json:"clientID,omitempty"`
-	// +kubebuilder:validation:MinLength=2
 	// +kubebuilder:validation:Required
 	// The shared oidc secret
-	ClientSecret string `json:"clientSecret,omitempty"`
+	ClientSecretRef corev1.ObjectReference `json:"clientSecretRef,omitempty"`
```
Comment (robinbobbitt) [inline api/v1alpha1/dexclient_types.go:34]: Per @itdove I believe this should be a SecretReference.

File: controllers/dexclient_controller.go [hunk 5]
```diff
@@ -106,7 +111,7 @@ func (r *DexClientReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 			dexv1Client.Spec.ClientID, // name
 			dexv1Client.Name,          // clientid
 			dexv1Client.Spec.LogoURL,
-			dexv1Client.Spec.ClientSecret, // secret
+			dexclientclientSecret, // secret
```
Comment (robinbobbitt) [inline controllers/dexclient_controller.go:114]: Do we need the same change for UpdateClient? Will this handle the case where the secret changes?

**Review by robinbobbitt — COMMENTED** [review]
Nice changes. I've added a couple more comments.

File: controllers/dexclient_controller.go (hunk 5 above)
Comment (itdove) [reply controllers/dexclient_controller.go:114]: You will need to add a watcher on the secret in order to regenerate the dex server configuration.

File: hack/generate-cr.sh (hunk 1 above)
Comment (cdoan1) [reply hack/generate-cr.sh]: yes, updated my script to use this now

File: hack/generate-cr.sh [hunk 6]
```diff
@@ -109,9 +134,9 @@ spec:
             - preferred_username
             - email
             - name
-        clientID: ${CLIENT_NAME}
```
Comment (vidyanambiar) [inline hack/generate-cr.sh:112]: @cdoan1 I believe this should point to DEXCLIENT_ID instead of DEXCLIENT_NAME

File: hack/generate-cr.sh (hunk 2 above)
Comment (cdoan1) [reply hack/generate-cr.sh:71]: Updated

File: hack/generate-cr.sh (hunk 3 above)
Comment (cdoan1) [reply hack/generate-cr.sh:110]: yeah, right now, whats in the generate-cr.sh will work. 
next PR, I'll swap the api code to match name = name, and id = id
I guess the other project wanted to make the id == name.

File: controllers/dexclient_controller.go (hunk 5 above)
Comment (cdoan1) [reply controllers/dexclient_controller.go:114]: we only set the dexclient secret on create
we don't update the secret, we can update all other bits but not the secret

File: hack/generate-cr.sh (hunk 3 above)
Comment (cdoan1) [reply hack/generate-cr.sh:110]: actually, i swap this now to what you guys expect

**Review by vidyanambiar — APPROVED** [review]

**Comment by vidyanambiar** [comment]
Thanks for the updates, Chris! This is working well for me now.

**Pull request merged**